tests/cpp/test_reduction.cpp

```diff
@@ -2627,4 +2627,146 @@ TEST_F(ReductionTest, TensorRankLimit) {
+  [[maybe_unused]] auto dtype = DataType::Float;
+  [[maybe_unused]] int64_t dtype_bytes = dataTypeSizeByte(dtype);
+  [[maybe_unused]] auto [ndims, inner_size] = GetParam();
+
+  std::vector<int64_t> element_at_each_dim;
+  if (ndims == 2) {
+    element_at_each_dim = {64, inner_size};
+  } else {
+    element_at_each_dim = {64, 8, inner_size};
+  }
+
+  std::vector<int64_t> reduced_dims;
+  int64_t reduced_elem_count = 1;
+  for (int i = 1; i < ndims; ++i) {
+    reduced_dims.push_back(i);
+    reduced_elem_count *= element_at_each_dim[i];
+  }
+
+  if (reduced_elem_count * dtype_bytes % 16 != 0) {
+    GTEST_SKIP() << "Reduced bytes is not divisible by 16, can't use TMA, "
+                    "reduced_elem_count: "
+                 << reduced_elem_count << ", dtype_bytes: " << dtype_bytes;
+    return;
+  }
+
+  const int64_t min_inner_dim = 2 * 16 / dtype_bytes;
+  if (reduced_elem_count % min_inner_dim != 0) {
+    GTEST_SKIP()
+        << "Reduced elements is not divisible by min_inner_dim, can't use TMA, "
+           "reduced_elem_count: "
+        << reduced_elem_count << ", min_inner_dim: " << min_inner_dim;
+    return;
+  }
+
+  auto fusion_ptr = std::make_unique<Fusion>();
+  FusionGuard fg(fusion_ptr.get());
+  Fusion& fusion = *fusion_ptr;
+
+  auto tv0 = makeContigTensor(ndims);
+  fusion.addInput(tv0);
+  auto tv1 = sum(tv0, reduced_dims);
+  fusion.addOutput(tv1);
+
+  auto options = at::TensorOptions().device(at::kCUDA, 0);
+  at::Tensor t0 = at::randn(element_at_each_dim, options);
+
+  // Merge the contiguous inner reduction dimensions
+  for (int64_t i = tv1->nDims() - 2; i >= 0; --i) {
+    if (tv1->axis(i)->isReduction() && tv1->axis(i + 1)->isReduction()) {
+      tv1->merge(i, i + 1);
+    }
+  }
+  // TODO: Split the inner dimension to ensure the TMA fits into smem.
+
+  // Propagate the merge to tv0
+  TransformPropagator propagator(tv1);
+  MaxLogicalDomainInfoSpanningTree(tv1).traverse(&propagator);
+
+  // Setup the TMA load. No TMA store.
+  auto tv0smem = tv0->cacheAfter(LoadStoreOpType::CpAsyncBulk);
+  tv0smem->setMemoryType(MemoryType::Shared);
+
+  tv0smem->axis(0)->parallelize(ParallelType::BIDx);
+  tv0smem->axis(1)->parallelize(ParallelType::Bulk);
+
+  auto tv1regs = tv1->cacheBefore();
+  auto redu_tv = tv1regs;
+
+  const int64_t vect = 4;
+  const int64_t tidx = 256;
+
+#define split_and_parallelize(tv, dim, size, parallelType) \
+  tv->split(dim, size);                                    \
+  tv->axis(dim + 1)->parallelize(parallelType);
+
+  // clang-format off
+  split_and_parallelize(redu_tv, 1, vect, ParallelType::Serial);   // 5
+  split_and_parallelize(redu_tv, 1, tidx, ParallelType::TIDx);     // 4
+  split_and_parallelize(redu_tv, 1,    4, ParallelType::Unroll);   // 3
+  split_and_parallelize(redu_tv, 1,    1, ParallelType::Unswitch); // 2
+
+  redu_tv->axis(1)->parallelize(ParallelType::Serial);             // 1
+  redu_tv->axis(0)->parallelize(ParallelType::BIDx);               // 0
```

Comment:
Why do we need to BID parallelize `tv0smem`, `redu_tv` and `tv1`, but not `tv0`?

The parallelization from `tv0smem` seems to suffice for `tv0`. But The same isn't true for `tv1` getting its parallelization from `redu_tv`. Is there a different semantic between `cacheAfter()` and `cacheBefore()`?

Reply:
>  Why do we need to BID parallelize tv0smem, redu_tv and tv1, but not tv0?

`tv0` is input, we don't need to compute it, so we don't need to parallelize it.
Saying a TensorView is “transformed/parallelized” means we are instructing it to compute all of its elements according to the parallelization strategy applied to it